### ISSUE TITLE
torch.fmod and torch.remainder return NaN on the vectorized path where the scalar path is correct (the fmod half of #77742)

### ISSUE THREAD

**Issue by krrishapatel** [pytorch/pytorch#193753]
### 🐛 Describe the bug

`torch.fmod` and `torch.remainder` return NaN when `|a / b|` overflows, but only for tensors long enough to reach the vectorized kernel. The same values in a shorter tensor give the correct answer.

```python
import torch, math

for n in (3, 4):
    x = torch.full((n,), 1e308, dtype=torch.float64)
    y = torch.full((n,), 0.5, dtype=torch.float64)
    print(n, torch.fmod(x, y)[0].item(), torch.remainder(x, y)[0].item())
# 3 0.0 0.0
# 4 nan nan

print(math.fmod(1e308, 0.5))  # 0.0
```

float32 is the same, with the threshold at 8 elements:

```python
x = torch.full((8,), 3.4028235e38, dtype=torch.float32)
y = torch.full((8,), -1.4e-45, dtype=torch.float32)
print(torch.remainder(x, y)[0].item())  # nan
print(torch.remainder(x[:1], y[:1])[0].item())  # 0.0
# numpy and Python both give -0.0
```

This is the case @peterbell10 posted in #77742. That issue was closed by #129536, which fixed `div` with `rounding_mode="floor"` by working around the bad `fmod` result inside `div_floor_floating_vec`:

https://github.com/pytorch/pytorch/blob/04b971abedf3167a3f929e4e7a9a5b7e8647c02c/aten/src/ATen/native/cpu/BinaryOpsKernel.cpp#L261-L263

`fmod_kernel` and `remainder_kernel` call `Vectorized::fmod` with no such fixup, so the `torch.fmod` half of #77742 is still there:

- `fmod_kernel`: https://github.com/pytorch/pytorch/blob/04b971abedf3167a3f929e4e7a9a5b7e8647c02c/aten/src/ATen/native/cpu/BinaryOpsKernel.cpp#L1056
- `remainder_kernel`: https://github.com/pytorch/pytorch/blob/04b971abedf3167a3f929e4e7a9a5b7e8647c02c/aten/src/ATen/native/cpu/BinaryOpsKernel.cpp#L404
- `remainder_kernel`, bf16 path: https://github.com/pytorch/pytorch/blob/04b971abedf3167a3f929e4e7a9a5b7e8647c02c/aten/src/ATen/native/cpu/BinaryOpsKernel.cpp#L382

The cause is the one already identified in #77742: `Sleef_fmod` is documented as undefined once `|x / y|` reaches 1e300 for double or 1e38 for float. Every vec backend implements `Vectorized<T>::fmod` as a direct Sleef call, so this is not specific to one ISA.

I know results are not guaranteed bitwise identical across vectorization. This is not a rounding difference. It is NaN in place of a finite value, and it flips based on tensor length alone.

One thing to watch in a fix: the result is not always zero when the division overflows, so the value cannot just be replaced with 0.

```python
math.fmod(-1.7976931348623157e308, 1.5e-323)  # -1e-323
```

Using the scalar `std::fmod` for the lanes where `(a / b).abs()` is infinite and `a` is finite would handle it, and that mask is already being computed in `div_floor_floating_vec`. Glad to send a PR if that is the direction you want.

Found by a differential tester that varies tensor length while holding the values fixed.

### Versions

torch 2.13.0, CPU, macOS arm64.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01

**Comment by krrishapatel** [comment]
Sent #193763.

Also hits bfloat16 `remainder`, since that kernel converts to float and bfloat16 has float32's exponent range. `fmod` is fine for bfloat16 and half because `Vectorized` implements it with `std::fmod` per element.

**Comment by krrishapatel** [comment]
Another repro that does not need extreme values, just a subnormal divisor:

```python
import torch, numpy as np, struct
tiny = struct.unpack("f", struct.pack("I", 0x00010000))[0]  # 9.18e-41

for n in (4, 8):
    a = torch.full((n,), 0.0314941, dtype=torch.float32)
    b = torch.full((n,), tiny, dtype=torch.float32)
    print(n, torch.fmod(a, b)[0].item(), torch.remainder(a, b)[0].item())
# 4 0.0 0.0
# 8 nan nan

print(np.fmod(0.0314941, tiny))  # 0.0
```

Same cause, `abs(a / b)` is 3.4e38 so it passes the 1e38 Sleef bound, but neither input is unusual. float64 does it too with a subnormal divisor.

Also worth noting for scope: `Vectorized<BFloat16>::fmod` hits this whenever it widens. Over all 2^32 bfloat16 pairs, widening to float differs from `std::fmod` on 528480772 of them, and every single one has `abs(a / b) >= 1e38`. Nothing inside the documented range differs. Half can never reach the bound at all, its largest possible ratio is about 1e12, so it is safe to widen (#193780).
